gsplat/cuda/_wrapper.py

```diff
@@ -0,0 +1,1120 @@
+
+
+def projection(
+    means: Tensor,  # [N, 3]
+    covars: Tensor,  # [N, 6] or None
+    quats: Tensor,  # [N, 4] or None
+    scales: Tensor,  # [N, 3] or None
+    viewmats: Tensor,  # [C, 4, 4]
+    Ks: Tensor,  # [C, 3, 3]
+    width: int,
+    height: int,
+    eps2d: float = 0.3,
+    near_plane: float = 0.01,
+    far_plane: float = 1e10,
+    radius_clip: float = 0.0,
+    packed: bool = False,
+    sparse_grad: bool = False,
+    calc_compensations: bool = False,
+) -> Tuple[Tensor, Tensor, Tensor, Tensor, Tensor]:
+    """Projects Gaussians to 2D.
+
+    .. note::
+
+        During projection, we ignore the Gaussians that are outside of the camera frustum.
+        So not all the elements in the output tensors are valid. The output `radii` could serve as
+        an indicator, in which zero radii means the corresponding elements are invalid in
+        the output tensors and will be ignored in the next rasterization process. If `packed=True`,
+        the output tensors will be packed into a flattened tensor, in which all elements are valid.
+        In this case, a `rindices` tensor and `cindices` tensor will be returned to indicate the
+        row (camera) and column (Gaussian) indices of the packed flattened tensor, which is essentially
+        following the COO sparse tensor format.
+
+    .. note::
+
+        This functions supports projecting Gaussians with either covariances or {quaternions, scales},
+        which will be converted to covariances internally in a fused CUDA kernel. Either `covars` or
+        {`quats`, `scales`} should be provided.
+
+    Args:
+        means: Gaussian means. [N, 3]
+        covars: Gaussian covariances (flattened upper triangle). [N, 6] Optional.
+        quats: Quaternions (No need to be normalized). [N, 4] Optional.
+        scales: Scales. [N, 3] Optional.
+        viewmats: Camera-to-world matrices. [C, 4, 4]
+        Ks: Camera intrinsics. [C, 3, 3]
+        width: Image width.
+        height: Image height.
+        eps2d: A epsilon added to the 2D covariance for numerical stability. Default: 0.3.
+        near_plane: Near plane distance. Default: 0.01.
+        far_plane: Far plane distance. Default: 1e10.
+        radius_clip: Gaussians with projected radii smaller than this value will be ignored. Default: 0.0.
+        packed: If True, the output tensors will be packed into a flattened tensor. Default: False.
+        sparse_grad: This is only effective when `packed` is True. If True, during backward the gradients
+          of {`means`, `covars`, `quats`, `scales`} will be a sparse Tensor in COO layout. Default: False.
+        calc_compensations: If True, a view-dependent opacity compensation factor will be computed, which
+          is useful for anti-aliasing. Default: False.
+
+    Returns:
+        A tuple:
+
+        If `packed` is True:
+
+        - **rindices**. The row indices of the projected Gaussians. Int32 tensor of shape [nnz].
```

Comment:
I think this needs to be renamed to be clearer, row/column being camera and gaussian idcs right?

Reply:
done! `rindices` -> `camera_ids`, `cindices` -> `gaussian_ids`